File: Scala/01_algorithmic_toolbox/src/main/scala-2.12/warmup/MaximumPairwiseProduct.scala
package warmup

object MaximumPairwiseProduct {
  final case class LargestElems(largest: Int, secondLargest: Int)

  def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  def calcMaximumPairwiseProduct(list: List[Int]): Long = {
    def processNextElem(acc: LargestElems, elem: Int): LargestElems =
      if (elem > acc.largest) LargestElems(largest = elem, secondLargest = acc.largest)
      else if (elem > acc.secondLargest) acc.copy(secondLargest = elem)
      else acc

    val largestElems: LargestElems = list.foldLeft(LargestElems(Int.MinValue, Int.MinValue))(processNextElem)
    largestElems.largest.toLong * largestElems.secondLargest.toLong
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val list: List[Int] = convertToIntList(reader.next())
    val result: Long = calcMaximumPairwiseProduct(list)
    println(result)
  }
}

File: OCaml/algorithmic_toolbox/warmup/maximum_pairwise_product.ml
type largest_elems = {largest : int; second_largest : int}


let convert_to_inlist line = List.map int_of_string (Str.split (Str.regexp " ") line)


let process_next_elem acc x =
    if x > acc.largest then {largest = x; second_largest = acc.largest}
    else if x > acc.second_largest then {acc with second_largest = x}
    else acc


let calc_maximum_pairwise_product list =
    let result = List.fold_left process_next_elem {largest = min_int; second_largest = min_int} list in
    result.largest * result.second_largest


let () =
    let line = read_line () in
    let list = convert_to_inlist line in
    print_int (calc_maximum_pairwise_product list);
    print_newline ()
